### PR TITLE
Add support for sampling rate in `streamlog`

### PR DESCRIPTION
**Pull request by timvaillancourt**
## Description

This PR implements the RFC https://github.com/vitessio/vitess/issues/15909 by adding a flag `--querylog-sample-rate float` to `vtcombo`, `vtgate` and `vttablet` in order for queries to be sampled randomly without the need for the client to trigger query logging via a query directive/comment == `--querylog-filter-tag string`

This flag supports values between `0.0` (no logging) and `1.0` (log all queries) to match other `0.0-1.0` float "sample" flags

`math/rand/v2` was used because it sounds better

## Related Issue(s)

Resolves https://github.com/vitessio/vitess/issues/15909

## Checklist

-   [ ] "Backport to:" labels have been added if this change should be back-ported to release branches
-   [ ] If this change is to be back-ported to previous releases, a justification is included in the PR description
-   [ ] Tests were added or are not required
-   [ ] Did the new or modified tests pass consistently locally and on CI?
-   [ ] Documentation was added or is not required

### CODE REVIEW

**Comment by deepthi** [comment]
Flag docs are auto-generated, so we don't need any manual changes on the website for this. @frouioui can you confirm that I'm getting this right?

**Comment by frouioui** [comment]
> Flag docs are auto-generated, so we don't need any manual changes on the website for this. @frouioui can you confirm that I'm getting this right?

@deepthi according https://github.com/vitessio/vitess-bot/blob/main/README.md#vitess-bot it should generate a PR upon merge to `main` and upon the next release. If it does not, I will investigate and in the meantime flags docs can be manually generated using the tool.

**Comment by frouioui** [comment]
After talking with @mattlord, I realized that actually the automation does not work.

**Review by deepthi — APPROVED** [review]

**Comment by deepthi** [comment]
> After talking with @mattlord, I realized that actually the automation does not work. We should update the website docs manually at the same time as this PR.

@frouioui can you provide the instructions for that? Or should we simply plan to do that manually for v20 before release? I suspect there may be other PRs which are missing website documentation as well. One example: #16021 
cc @shlomi-noach for v20 release decision.

**Comment by frouioui** [comment]
Yeah exactly running `VITESS_DIR=../vitess ./tools/sync_cobradocs.sh` from inside the website repo works fine too to fill all the missed changes.

**Review by frouioui — APPROVED** [review]
LGTM

**Comment by timvaillancourt** [comment]
@deepthi / @frouioui, here are the metric outputs:
```
tvaillancourt@REDACTED:~$ curl -s localhost:15000/debug/vars | jq .QuerySQLTextCounts
{
  "Join.Select.": 3234,
  "dual.Select.": 3615
}
tvaillancourt@REDACTED:~$ curl -s localhost:15000/metrics 2>&1 | grep vttablet_query_sql_text_counts
# HELP vttablet_query_sql_text_counts query sql text counts
# TYPE vttablet_query_sql_text_counts counter
vttablet_query_sql_text_counts{plan="Select",table="Join",workload=""} 3234
vttablet_query_sql_text_counts{plan="Select",table="dual",workload=""} 3650
```

So looks good 👍. While doing that I thought of two potential alternatives in case this didn't work: `QueryTextProcessed` or `QueryTextRead`, any thoughts?

**Comment by timvaillancourt** [comment]
To address my own paranoia regarding this thread I added a go benchmark:
```
BenchmarkShouldEmitLog
BenchmarkShouldEmitLog/default
BenchmarkShouldEmitLog/default-12         	327598167	         3.739 ns/op
BenchmarkShouldEmitLog/filter_tag
BenchmarkShouldEmitLog/filter_tag-12      	100000000	        11.67 ns/op
BenchmarkShouldEmitLog/50%_sample_rate
BenchmarkShouldEmitLog/50%_sample_rate-12 	79597881	        15.05 ns/op
```
Where `default` = no filter-tag or random saming, `filter_tag` = query-comment filter tag, `50%_sample_rate` is `--query-sample-rate 0.5`

I think I'm most surprised by `filter_tag` which essentially adds a `string.Contains(sql)` actually :D

**Comment by shlomi-noach** [comment]
Re docs, good points. My suggestion: after code freeze, I'll generate `vtctldclient` docs manually. At any case we can decouple the docs discussion from this PR.

File: go/streamlog/streamlog_test.go [hunk 1]
```diff
@@ -264,18 +265,39 @@ func TestFile(t *testing.T) {
...
 }
 
+func TestShouldSampleQuery(t *testing.T) {
+	queryLogSampleRate = -1
+	assert.False(t, shouldSampleQuery())
+
+	queryLogSampleRate = 0
+	assert.False(t, shouldSampleQuery())
+
+	// for test coverage, can't test a random result
```
Comment (shlomi-noach) [inline go/streamlog/streamlog_test.go:275]: You can run 100 consecutive randoms and expect _some_ to pass and _some_ to fail. The false negative case for 1 in 2^100 is lower than any computational problem in modern hardware.

**Review by shlomi-noach — APPROVED** [review]

**Pull request merged**